lambdas/src/apis/collections/controllers/collections.ts

```diff
@@ -150,46 +211,132 @@ function buildUrn(protocol: string, contract: string, option: string): string {
-  } catch (e) {
-    res.status(500).send(e.message)
-  }
+function validateSize(size: string = DEFAULT_IMAGE_SIZE): size is ValidSize {
+  if (!isValidSize(size)) throw new ServiceError('Invalid size')
+  return true
+}
+
+function getFileHash(entity: Entity, fileName?: string): string {
+  const hash = findHashForFile(entity, fileName)
+  if (!hash) throw new ServiceError(`Hash not found for file ${fileName}`, 404)
+
+  return hash
 }
 
-async function fetchEntity(client: SmartContentClient, urn: string): Promise<Entity | undefined> {
+async function fetchEntity(client: SmartContentClient, urn: string): Promise<Entity> {
   const entities: Entity[] = await client.fetchEntitiesByPointers(EntityType.WEARABLE, [urn])
-  return entities && entities.length > 0 && entities[0].metadata ? entities[0] : undefined
+  if (!(entities && entities.length > 0 && entities[0].metadata)) throw new ServiceError('Entity not found', 404)
+
+  return entities[0]
 }
 
-export function getBodyShapes(representations: WearableMetadataRepresentation[]) {
+function getBodyShapes(representations: WearableMetadataRepresentation[]): string[] {
   const bodyShapes = new Set<WearableBodyShape>()
   for (const representation of representations) {
     for (const bodyShape of representation.bodyShapes) {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
       bodyShapes.add(bodyShape.split(':').pop()!)
     }
   }
   return Array.from(bodyShapes)
 }
 
+function isValidSize(size: string): size is ValidSize {
+  return sizes[size] !== undefined
+}
+
+function getImagePath(root: string, imageRequest: ImageRequest): string {
+  return root + `/images/` + getImageRequestId(imageRequest) + '.png'
+}
+
+// Using this folder structure allow us to find and remove older versions of the same urn (entity)
+function getImageRequestId({ urn, hash, size, rarityBackground }: ImageRequest): string {
+  return `${urn}/${hash}/${rarityBackground ?? 'thumbnail'}-${size}`
+}
+
+// Delete all images that are not the latest version (same hash)
+async function pruneObsoleteImages(root: string, urn: string, hash: string) {
+  const existsFolder = await checkFileExists(root + `/images/` + urn + '/' + hash)
+  if (!existsFolder) await cleanFolder(root + `/images/` + urn)
+}
+
+async function cleanFolder(folderPath: string) {
+  await fs.promises.rm(folderPath, { recursive: true, force: true })
+}
+
+async function getImage(
+  client: SmartContentClient,
+  rootStorageLocation: string,
+  imageRequest: ImageRequest
+): Promise<[NodeJS.ReadableStream, number]> {
+  const path = getImagePath(rootStorageLocation, imageRequest)
+
+  // Check if the image is already in the cache, otherwise build and store it
+  if (await checkFileExists(path)) {
+    metricsComponent.increment('cache_hit', { image_size: imageRequest.size })
+  } else {
+    metricsComponent.increment('cache_miss', { image_size: imageRequest.size })
+    await saveImage(client, rootStorageLocation, imageRequest)
+
```

Comment:
what do you think of adding the prune folder here? So we don't have to check twice if the folder exists

Reply:
I think is not the case here, as we are checking for image file and not hash folder. It should be possible that no prune should be done at this point.